### PR TITLE
Finalize eligibility check for augmenting visualizations by vis augmenter

### LINKED ISSUE #3268: [Feature Anywhere] Finalize eligibility based on vis configuration
As part of the logic for rendering plugin resources & overlays on a visualization, we need to have a clean way for determining a particular vis config is eligible for the feature.

Eligibility checks will be needed when:
- collecting `VisLayer`s or not in the `VisualizeEmbeddable` render flow
- displaying the extra options in the context menu (the plugin-specific options, the 'View Events' option) - this will be done via `isCompatible()` fns of the actions that are triggered via `CONTEXT_MENU_TRIGGER`.

This roughly will include:
- must be of type 'line'
- must have a "valid" config (one set of x axes, x axis field is of type `date`, reasonable metrics, etc.). this will be the bulk of the work needed
- must not show in the 'edit' context. basically, when viewing the details, we don't want to show, or even attempt to show, any VisLayers here. the feature is just for showing in the context of the dashboard. later improvements to the feature may support something to show / configure in the edit view

### PR DESCRIPTION
**Pull request by lezzago**
### Description
Finalize eligibility check for augmenting visualizations by vis augmenter

**Eligibility Requirements:**
- Must be a vislib line chart type
- Must contain one Y-axis metric aggregation
- Must not have non Y-axis `metric` aggregation types
- Must use `date histogram` aggregation type for the X-axis bucket
- Must have X-axis on the bottom
- There can only be 1 X-axis aggregation bucket setup
- The dimensions must be valid as well, this is determined by having an X-axis to be setup
 
### Issues Resolved
#3268
 
### Check List
- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/vis_augmenter/public/utils/utils.test.ts [hunk 1]
```diff
@@ -16,26 +16,150 @@ import {
...
+        },
+        data: {
+          aggs,
         },
       } as unknown) as Vis;
-      expect(isEligibleForVisLayers(vis)).toEqual(true);
+      expect(isEligibleForVisLayers(vis, validDimensions)).toEqual(false);
+    });
+    it('vis is eligible with valid type', async () => {
+      expect(isEligibleForVisLayers(validVis, validDimensions)).toEqual(true);
```
Comment (ohltyler) [inline src/plugins/vis_augmenter/public/utils/utils.test.ts:162]: Can you add a test for invalid dimensions?

File: src/plugins/vis_augmenter/public/utils/utils.ts [hunk 2]
```diff
@@ -13,10 +13,15 @@ import {
 } from '../../../../plugins/expressions/public';
 import { ISavedAugmentVis, SavedAugmentVisLoader, VisLayerFunctionDefinition } from '../';
 
-// TODO: provide a deeper eligibility check.
-// Tracked in https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3268
-export const isEligibleForVisLayers = (vis: Vis): boolean => {
-  return vis.params.type === 'line';
+export const isEligibleForVisLayers = (vis: Vis, dimensions: VislibDimensions): boolean => {
+  const isDateHistogram =
+    vis.data.aggs?.byTypeName('date_histogram').length === 1 && vis.data.aggs?.aggs.length === 2;
```
Comment (ohltyler) [inline src/plugins/vis_augmenter/public/utils/utils.ts:18]: Can you explain what `vis.data.aggs?.aggs.length === 2` is doing? adding a comment in the code might be worthwhile too

File: src/plugins/vis_augmenter/public/utils/utils.test.ts [hunk 3]
```diff
@@ -16,26 +16,150 @@ import {
...
+        enabled: true,
+        type: 'histogram',
+        params: {},
+      },
+      {
+        enabled: true,
+        type: 'date_histogram',
+        params: {},
+      },
+    ];
```
Comment (ohltyler) [inline src/plugins/vis_augmenter/public/utils/utils.test.ts:48]: why is there both a `histogram` and a `date_histogram` here?

File: src/plugins/vis_augmenter/public/utils/utils.ts [hunk 4]
```diff
@@ -13,10 +13,15 @@ import {
...
+  let isOnlyLine = vis.params.type === 'line';
+  vis.params.seriesParams.forEach((seriesParam: { type: string }) => {
+    isOnlyLine = isOnlyLine && seriesParam.type === 'line';
+  });
+  const isValidDimensions = dimensions.x !== null;
```
Comment (ohltyler) [inline src/plugins/vis_augmenter/public/utils/utils.ts:23]: What about the `position` needing to be on `bottom`?

File: src/plugins/vis_type_vega/public/expressions/line_vega_spec_fn.ts [hunk 5]
```diff
@@ -68,7 +69,22 @@ export const formatDataTable = (
   datatable.columns.forEach((column) => {
     // clean quotation marks from names in columns
     column.name = cleanString(column.name);
+    // clean ids to remove "." as that will cause vega to not process it correctly.
+    // This happens for different metric types
+    column.id = column.id.replaceAll('.', '-');
   });
+
+  // clean row keys to remove "." as that will cause vega to not process it correctly
```
Comment (ohltyler) [inline src/plugins/vis_type_vega/public/expressions/line_vega_spec_fn.ts:77]: I think this is worth adding some tests for some of these scenarios & edge cases

File: src/plugins/vis_type_vega/public/expressions/line_vega_spec_fn.ts [hunk 6]
```diff
@@ -133,14 +149,16 @@ export const buildYAxis = (
...
   visParams: VisParams
 ) => {
+  const subAxis = {
+    title: cleanString(valueAxis.title.text) || column.name,
+    grid: visParams.grid.valueAxis !== undefined,
+    orient: valueAxis.position,
+    labels: valueAxis.labels.show,
+    labelAngle: valueAxis.labels.rotate,
+  };
+  if (column.meta?.type === 'percentile_ranks') Object.assign(subAxis, { format: '.0%' });
```
Comment (ohltyler) [inline src/plugins/vis_type_vega/public/expressions/line_vega_spec_fn.ts:159]: Can you add a comment about why this is needed? are there other types that could potentially need other formatting too?

File: src/plugins/vis_type_vislib/opensearch_dashboards.json [hunk 7]
```diff
@@ -3,7 +3,7 @@
   "version": "opensearchDashboards",
   "server": true,
   "ui": true,
-  "requiredPlugins": ["charts", "data", "expressions", "visualizations", "opensearchDashboardsLegacy", "visTypeVega"],
+  "requiredPlugins": ["charts", "data", "expressions", "visualizations", "opensearchDashboardsLegacy", "visTypeVega", "visAugmenter"],
   "optionalPlugins": ["visTypeXy"],
   "requiredBundles": ["opensearchDashboardsUtils", "visDefaultEditor"]
```
Comment (ohltyler) [inline src/plugins/vis_type_vislib/opensearch_dashboards.json:8]: If the helper fn src code is the only part needed from `visAugmenter`, prefer to add into `requiredBundles` instead of `requiredPlugins`. `requiredPlugins` is more reserved for accessing services needed from the parent plugin when running the `setup()`/`start()` lifecycle fns.

File: src/plugins/vis_type_vislib/public/line_to_expression.ts [hunk 8]
```diff
@@ -26,9 +27,13 @@ export const toExpressionAst = async (vis: Vis, params: any) => {
...
 
   // Checks if there are vislayers to overlay. If not, default to the vislib implementation.
-  if (params.visLayers == null || Object.keys(params.visLayers).length === 0) {
+  const dimensions: VislibDimensions = await buildVislibDimensions(vis, params);
+  if (
+    !isEligibleForVisLayers(vis, dimensions) ||
+    params.visLayers == null ||
+    Object.keys(params.visLayers).length === 0
+  ) {
     // This wont work but is needed so then it will default to the original vis lib renderer
```
Comment (ohltyler) [inline src/plugins/vis_type_vislib/public/line_to_expression.ts:36]: nit: is this comment outdated? not sure what it means exactly

Comment (lezzago) [reply src/plugins/vis_type_vislib/public/line_to_expression.ts:36]: To automatically default back to the non-expression method (render using vega-lite), the `toExpressionAST` function needs to fail since in the line definition, we cannot logically choose one over the other.

Comment (lezzago) [reply src/plugins/vis_type_vislib/public/line_to_expression.ts:36]: I will create a github issue to look more into this to see what needs to change for it to have a better switching mechanism.

File: src/plugins/vis_augmenter/public/utils/utils.ts (hunk 4 above)
Comment (lezzago) [reply src/plugins/vis_augmenter/public/utils/utils.ts:32]: done

File: src/plugins/vis_augmenter/public/utils/utils.test.ts (hunk 3 above)
Comment (lezzago) [reply src/plugins/vis_augmenter/public/utils/utils.test.ts]: Removed histogram and used other types that made more sense.

File: src/plugins/vis_type_vega/public/expressions/line_vega_spec_fn.ts (hunk 5 above)
Comment (lezzago) [reply src/plugins/vis_type_vega/public/expressions/line_vega_spec_fn.ts]: Added tests

File: src/plugins/vis_type_vislib/opensearch_dashboards.json (hunk 7 above)
Comment (lezzago) [reply src/plugins/vis_type_vislib/opensearch_dashboards.json]: done

File: src/plugins/vis_augmenter/public/utils/utils.test.ts (hunk 1 above)
Comment (lezzago) [reply src/plugins/vis_augmenter/public/utils/utils.test.ts:224]: done

File: src/plugins/vis_augmenter/public/utils/utils.test.ts [hunk 9]
```diff
@@ -16,26 +16,150 @@ import {
   getMockAugmentVisSavedObjectClient,
   generateAugmentVisSavedObject,
 } from '../saved_augment_vis';
+import { AggConfigs, AggTypesRegistryStart } from '../../../data/common/search/aggs';
+import { stubIndexPatternWithFields } from '../../../data/common/index_patterns/index_pattern.stub';
+import { IndexPattern } from '../../../data/common/index_patterns/index_patterns';
+import { mockAggTypesRegistry } from '../../../data/common/search/aggs/test_helpers';
```
[the remark this answers is not in the document]
Comment (lezzago) [reply src/plugins/vis_augmenter/public/utils/utils.test.ts]: done

File: src/plugins/vis_type_vega/public/expressions/line_vega_spec_fn.ts (hunk 6 above)
Comment (lezzago) [reply src/plugins/vis_type_vega/public/expressions/line_vega_spec_fn.ts]: done

File: src/plugins/vis_type_vislib/public/line_to_expression.ts (hunk 8 above)
Comment (ohltyler) [reply src/plugins/vis_type_vislib/public/line_to_expression.ts:36]: It's not failing here though, rather it's just building the rest of the expression string without constructing it via the vega expression functions. I think we can just remove this comment to prevent confusion.

And overall, I think it's actually good design to have the entire string pipeline constructed in here (both vislib and vega implementations). Ideally, all vis types have an implemented `toExpressionAst`

File: src/plugins/vis_augmenter/public/utils/utils.ts [hunk 10]
```diff
@@ -14,14 +14,23 @@ import {
...
+  const isValidXaxis =
+    vis.data.aggs?.byTypeName('date_histogram').length === 1 &&
+    vis.params.categoryAxes.length === 1 &&
+    vis.params.categoryAxes[0].position === 'bottom';
+  // Support 1 segment for x axis bucket (that is date_histogram) and support metrics for
+  // multiple supported yaxis only. If there are other aggregation types, this is not
+  // valid for augmentation
+  const hasCorrectAggregationCount =
+    vis.data.aggs !== undefined &&
+    vis.data.aggs?.bySchemaName('metric').length === vis.data.aggs?.aggs.length - 1;
```
Comment (ohltyler) [inline src/plugins/vis_augmenter/public/utils/utils.ts:27]: Should we also check that there is >= 1 metric agg? Or should we still support 0?

File: src/plugins/vis_type_vega/public/expressions/helpers.test.js [hunk 11]
```diff
@@ -21,10 +21,25 @@ import {
...
+      );
+    });
+  });
+
   describe('cleanString()', function () {
-    it('string should not contain "', function () {
+    it('string should not contain', function () {
       const dirtyString = '"someString"';
       expect(cleanString(dirtyString)).toBe('someString');
     });
```
Comment (ohltyler) [inline src/plugins/vis_type_vega/public/expressions/helpers.test.js:45]: nit: should the test name still say `'string should not contain "'`?

Comment (ohltyler) [reply src/plugins/vis_type_vega/public/expressions/helpers.test.js:45]: or `string should not contain quotation marks`?

File: src/plugins/vis_augmenter/public/utils/utils.ts (hunk 10 above)
Comment (lezzago) [reply src/plugins/vis_augmenter/public/utils/utils.ts:27]: That is a fair point, there is no need to support a graph that will not have any metrics since there would be no reason to overlay events on the graph.

File: src/plugins/vis_type_vega/public/expressions/helpers.test.js (hunk 11 above)
Comment (lezzago) [reply src/plugins/vis_type_vega/public/expressions/helpers.test.js:45]: Good catch, I am changing it to quotation marks to ensure another dev doesn't accidentally remove `"` thinking it was a mistake.

**Review by ohltyler — APPROVED** [review]
LGTM!

File: src/plugins/vis_type_vislib/public/line_to_expression.ts (hunk 8 above)
Comment (lezzago) [reply src/plugins/vis_type_vislib/public/line_to_expression.ts]: Updated the comment. I checked again and it doesn't fail anymore and properly builds the vislib line chart

File: src/plugins/vis_augmenter/public/utils/utils.test.ts [hunk 12]
```diff
@@ -16,26 +16,234 @@ import {
...
+              position: 'top',
+            },
+          ],
+        },
+        data: {
+          aggs,
         },
       } as unknown) as Vis;
-      expect(isEligibleForVisLayers(vis)).toEqual(true);
+      expect(isEligibleForVisLayers(invalidVis, validDimensions)).toEqual(false);
```
Comment (abbyhu2000) [inline src/plugins/vis_augmenter/public/utils/utils.test.ts:243]: nice tests~

File: src/plugins/vis_augmenter/public/utils/utils.ts [hunk 13]
```diff
@@ -13,10 +13,25 @@ import {
...
+    vis.params.categoryAxes.length === 1 &&
+    vis.params.categoryAxes[0].position === 'bottom';
+  // Support 1 segment for x axis bucket (that is date_histogram) and support metrics for
+  // multiple supported yaxis only. If there are other aggregation types, this is not
+  // valid for augmentation
+  const hasCorrectAggregationCount =
+    vis.data.aggs !== undefined &&
+    vis.data.aggs?.bySchemaName('metric').length > 0 &&
+    vis.data.aggs?.bySchemaName('metric').length === vis.data.aggs?.aggs.length - 1;
+  let isOnlyLine = vis.params.type === 'line';
```
Comment (abbyhu2000) [inline src/plugins/vis_augmenter/public/utils/utils.ts:29]: just a nit: `isLine` might be sufficient?

File: src/plugins/vis_type_vega/public/expressions/helpers.ts [hunk 14]
```diff
@@ -49,7 +50,22 @@ export const formatDatatable = (
   datatable.columns.forEach((column) => {
     // clean quotation marks from names in columns
     column.name = cleanString(column.name);
+    // clean ids to remove "." as that will cause vega to not process it correctly.
```
Comment (abbyhu2000) [inline src/plugins/vis_type_vega/public/expressions/helpers.ts:53]: just a question here: can we just rename the ids so they do not include '.'?

**Review by abbyhu2000 — APPROVED** [review]
LGTM!

File: src/plugins/vis_type_vega/public/expressions/helpers.ts [hunk 15]
```diff
@@ -49,7 +50,22 @@ export const formatDatatable = (
...
+  const updatedRows: OpenSearchDashboardsDatatableRow[] = [];
+  datatable.rows.forEach((row) => {
+    const newRow = {};
+    for (const [key, value] of Object.entries(row)) {
+      const cleanKey = key.replaceAll('.', '-');
+      Object.assign(newRow, { [cleanKey]: value });
+    }
+    updatedRows.push(newRow);
+  });
+  datatable.rows = updatedRows;
```
Comment (joshuarrrr) [inline src/plugins/vis_type_vega/public/expressions/helpers.ts:68]: style nit suggestions - readability improvements are debatable, but you can avoid the temp variables with map and reduce.
```suggestion
  const updatedRows: OpenSearchDashboardsDatatableRow[] = datatable.rows.map((row) => (
    Object.entries(row).reduce((updatedRow, [key, value]) => {
      const cleanKey = key.replaceAll('.', '-');
      return Object.assign(updatedRow, { [cleanKey]: value });
    }, {});
  ));
  datatable.rows = updatedRows;
```

File: src/plugins/vis_type_vega/public/expressions/helpers.ts [hunk 16]
```diff
@@ -111,14 +127,17 @@ export const buildYAxis = (
...
 ) => {
+  const subAxis = {
+    title: cleanString(valueAxis.title.text) || column.name,
+    grid: visParams.grid.valueAxis !== undefined,
+    orient: valueAxis.position,
+    labels: valueAxis.labels.show,
+    labelAngle: valueAxis.labels.rotate,
+  };
+  // Percentile ranks aggregation metric needs percentile formatting.
+  if (column.meta?.type === 'percentile_ranks') Object.assign(subAxis, { format: '.0%' });
```
Comment (joshuarrrr) [inline src/plugins/vis_type_vega/public/expressions/helpers.ts:138]: interesting finding!

File: src/plugins/vis_type_vislib/public/line_to_expression.ts [hunk 17]
```diff
@@ -26,9 +27,13 @@ export const toExpressionAst = async (vis: Vis, params: any) => {
   );
 
   // Checks if there are vislayers to overlay. If not, default to the vislib implementation.
-  if (params.visLayers == null || Object.keys(params.visLayers).length === 0) {
-    // This wont work but is needed so then it will default to the original vis lib renderer
-    const dimensions = await buildVislibDimensions(vis, params);
+  const dimensions: VislibDimensions = await buildVislibDimensions(vis, params);
+  if (
+    !isEligibleForVisLayers(vis, dimensions) ||
```
Comment (joshuarrrr) [inline src/plugins/vis_type_vislib/public/line_to_expression.ts:32]: nit - should we check this last, after checking that there are visLayers?

File: src/plugins/vis_type_vislib/opensearch_dashboards.json [hunk 18]
```diff
@@ -5,5 +5,5 @@
   "ui": true,
   "requiredPlugins": ["charts", "data", "expressions", "visualizations", "opensearchDashboardsLegacy", "visTypeVega"],
   "optionalPlugins": ["visTypeXy"],
-  "requiredBundles": ["opensearchDashboardsUtils", "visDefaultEditor"]
+  "requiredBundles": ["opensearchDashboardsUtils", "visDefaultEditor", "visAugmenter"]
```
Comment (joshuarrrr) [inline src/plugins/vis_type_vislib/opensearch_dashboards.json:8]: In the future, I still think there's a better way to set this up, so that visAugmenter can simply specify that vega-lite rendering should be used, and it wouldn't be necessary to import here. But that's something we an come back to.

**Review by joshuarrrr — APPROVED** [review]
@lezzago Looks good - a couple questions and some minor style suggestions, which can be implemented in a follow-up PR if you'd prefer.

File: src/plugins/vis_augmenter/public/utils/utils.ts [hunk 19]
```diff
@@ -13,10 +13,25 @@ import {
...
-// TODO: provide a deeper eligibility check.
-// Tracked in https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3268
-export const isEligibleForVisLayers = (vis: Vis): boolean => {
-  return vis.params.type === 'line';
+export const isEligibleForVisLayers = (vis: Vis, dimensions: VislibDimensions): boolean => {
+  // Only support date histogram and ensure there is only 1 x-axis and it has to be on the bottom
+  const isValidXaxis =
+    vis.data.aggs?.byTypeName('date_histogram').length === 1 &&
+    vis.params.categoryAxes.length === 1 &&
+    vis.params.categoryAxes[0].position === 'bottom';
```
[the remark this answers is not in the document]
Comment (lezzago) [reply src/plugins/vis_augmenter/public/utils/utils.ts:21]: This is because we want to force it to be a bottom axis for the use case of the functionality. This would make it so it is aligned with the augmented triangles that indicate the alerts/anomalies.

[the remark this answers is not in the document]
Comment (joshuarrrr) [reply src/plugins/vis_augmenter/public/utils/utils.ts:21]: got it, makes sense. In the future, should be easy enough to allow it on top, too (put both the axis and the triangles up there), if that's something important to users.

File: src/plugins/vis_type_vega/public/expressions/helpers.ts (hunk 14 above)
Comment (lezzago) [reply src/plugins/vis_type_vega/public/expressions/helpers.ts:53]: If we change it how it is in `OpenSearchDashboardsDatatable`, this can cause a lot of implications as we need to modify `OpenSearchaggsExpressionFunction`.

File: src/plugins/vis_augmenter/public/utils/utils.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (lezzago) [reply src/plugins/vis_augmenter/public/utils/utils.ts:21]: We need to have it on the bottom since the augmentation from the alerts/anomalies would look odd when the axis is not on the bottom.

File: src/plugins/vis_type_vislib/opensearch_dashboards.json (hunk 18 above)
Comment (lezzago) [reply src/plugins/vis_type_vislib/opensearch_dashboards.json:8]: Yea that seems like a better way to structure this, but we would have to come back to that later.

File: src/plugins/vis_augmenter/public/utils/utils.test.ts [hunk 20]
```diff
@@ -16,26 +16,234 @@ import {
...
+          params: {},
+          schema: 'radius',
+        },
+        {
+          enabled: true,
+          type: 'metrics',
+          params: {},
+          schema: 'metrics',
+        },
+      ];
```
[the remark this answers is not in the document]
Comment (lezzago) [reply src/plugins/vis_augmenter/public/utils/utils.test.ts]: I made some updates for this to improve this, but more improvements can happen if we make these into data models. That can be done in a future PR.

**Review by joshuarrrr — APPROVED** [review]
